contracts/contracts/harvest/BaseHarvester.sol

```diff
@@ -99,87 +133,208 @@ abstract contract BaseHarvester is Governable {
-            allowedSlippageBps: _allowedSlippageBps,
-            harvestRewardBps: _harvestRewardBps,
-            uniswapV2CompatibleAddr: _uniswapV2CompatibleAddr,
-            doSwapRewardToken: _doSwapRewardToken,
-            liquidationLimit: _liquidationLimit
-        });
-
-        address oldUniswapAddress = rewardTokenConfigs[_tokenAddress]
-            .uniswapV2CompatibleAddr;
+        address oldRouterAddress = rewardTokenConfigs[_tokenAddress]
+            .swapRouterAddr;
         rewardTokenConfigs[_tokenAddress] = tokenConfig;
 
-        IERC20 token = IERC20(_tokenAddress);
-
-        address priceProvider = IVault(vaultAddress).priceProvider();
-
         // Revert if feed does not exist
         // slither-disable-next-line unused-return
-        IOracle(priceProvider).price(_tokenAddress);
+        IOracle(IVault(vaultAddress).priceProvider()).price(_tokenAddress);
 
+        IERC20 token = IERC20(_tokenAddress);
         // if changing token swap provider cancel existing allowance
         if (
-            /* oldUniswapAddress == address(0) when there is no pre-existing
+            /* oldRouterAddress == address(0) when there is no pre-existing
              * configuration for said rewards token
              */
-            oldUniswapAddress != address(0) &&
-            oldUniswapAddress != _uniswapV2CompatibleAddr
+            oldRouterAddress != address(0) &&
+            oldRouterAddress != newRouterAddress
         ) {
-            token.safeApprove(oldUniswapAddress, 0);
+            token.safeApprove(oldRouterAddress, 0);
+        }
+
+        // Give SwapRouter infinite approval when needed
+        if (oldRouterAddress != newRouterAddress) {
+            token.safeApprove(newRouterAddress, 0);
+            token.safeApprove(newRouterAddress, type(uint256).max);
         }
 
-        // Give Uniswap infinite approval when needed
-        if (oldUniswapAddress != _uniswapV2CompatibleAddr) {
-            token.safeApprove(_uniswapV2CompatibleAddr, 0);
-            token.safeApprove(_uniswapV2CompatibleAddr, type(uint256).max);
+        SwapPlatform _platform = tokenConfig.platform;
+        if (_platform == SwapPlatform.UniswapV2Compatible) {
+            uniswapV2Path[_tokenAddress] = _decodeUniswapV2Path(
+                swapData,
+                _tokenAddress
+            );
+        } else if (_platform == SwapPlatform.UniswapV3) {
+            uniswapV3Path[_tokenAddress] = _decodeUniswapV3Path(
+                swapData,
+                _tokenAddress
+            );
+        } else if (_platform == SwapPlatform.Balancer) {
+            balancerPoolId[_tokenAddress] = _decodeBalancerPoolId(swapData);
+        } else if (_platform == SwapPlatform.Curve) {
+            curvePoolIndices[_tokenAddress] = _decodeCurvePoolIndices(
+                swapData,
+                newRouterAddress,
+                _tokenAddress
+            );
+        } else {
+            revert("Unknown SwapPlatfrom");
```

Comment:
We could replace this with a solidity [custom error](https://ethereum.stackexchange.com/questions/101782/requirecondition-message-vs-revert-with-a-custom-error-which-is-better-a) that reduces the gas cost. @naddison36 has started doing this in [OETH Oracle PR](https://github.com/OriginProtocol/origin-dollar/pull/1815/files#diff-42dbcd9f1d9adfaa494e307193605f6ddf3ed0c17dbf0dfd03f5fc97514dc603R97) and I think it is a better approach. 

also Codecov is complaining this line isn't tested

Reply:
Good point about Custom Errors, Will change to use them. 

As for CodeCov, this line is unreachable since we revert in `setRewardTokenConfig` method if the `swapPlatform` is something else. I just added it here as a defence